Remove template specialization for toMsg functions

jacobperron opened the pull request:
Resolves #176 

Before, the compiler was preferring the overloaded function:

https://github.com/ros2/geometry2/blob/71b054249cda9f1bbdbdc2f61bd518b998ce0378/tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h#L397-L398

instead of the template specialization. I tried to change the overload into a template specialization, but then we run into the error that the compile can't deduce the second template parameter (ie. the return type).

I'm not sure if this is an ideal fix, or if instead we should go with template specialization (I'm not sure how we were/are benefiting from this), which requires the user to be explicit:

```c++
auto msg = tf2::toMsg<tf2::Stamped<tf2::Transform>, geometry_msgs::msg::TransformStamped>(tf_stamped);
```

I've also went ahead and made the change for stamped quaternions, which have the same bug.

jacobperron committed:
Add unit test for converting to/from stamped transform types

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Remove toMsg template specialization for stamped transforms

This fixes #176 as demonstrated by the unit test compiling.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Fix test

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Remove toMsg template specialization for stamped quaternions

Similar to stamped transforms, this resolves a compilation issue when using the function.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Expand test

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
Testing packages that depend on `tf2_geometry_msgs`:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=8322)](http://ci.ros2.org/job/ci_linux/8322/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=4264)](http://ci.ros2.org/job/ci_linux-aarch64/4264/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6785)](http://ci.ros2.org/job/ci_osx/6785/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=8222)](http://ci.ros2.org/job/ci_windows/8222/)

sloretz commented on tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h:
```diff
@@ -354,7 +354,6 @@ void fromMsg(const geometry_msgs::msg::QuaternionStamped& msg, geometry_msgs::ms
  */
-template <>
 inline
 geometry_msgs::msg::QuaternionStamped toMsg(const tf2::Stamped<tf2::Quaternion>& in)
```
Template specialization allows different return types for the same argument (though the template arguments must be given). I don't think these two could be in the same header as overloads, and maybe not even the same binary if the compiler ignores `inline`.

```c++
geometry::msgs::Quaternion toMsg(const tf2::Stamped<tf2::Quaternion>&);
geometry::msgs::QuaternionStamped toMsg(const tf2::Stamped<tf2::Quaternion>&)
```

That said, I didn't find anyone defining `QuaternionStamped` -> `Quaternion`.

jacobperron replied:
I would argue that the definition

```c++
geometry::msgs::Quaternion toMsg(const tf2::Stamped<tf2::Quaternion>&);
```

shouldn't be allowed since there's a loss of information, but perhaps there's valid use-cases for this.

IIUC, if we want to keep the template specialization, for this case to work with `tf2::convert` we should update [this line](https://github.com/ros2/geometry2/blob/71b054249cda9f1bbdbdc2f61bd518b998ce0378/tf2/include/tf2/impl/convert.h#L65) to:

```diff
-  b = toMsg(a);
+  b = toMsg<A, B>(a);
```

sloretz replied:
> I would argue that the definition shouldn't be allowed since there's a loss of information

Maybe so. I'm not sure what the motivation for template specialization is, but it seems like that my example is not a supported use case. If `class Q Stamped` converts to either `msg Q Stamped` or `msg Q`, and `msg Q Stamped` can be converted to `class Q Stamped`, then there would be a path to convert `msg Q Stamped` to `msg Q` via `msg Q Stamped -(fromMsg)> class Q Stamped -(toMsg)> msg Q`, but message to message conversions are forbidden by this comment.

https://github.com/ros2/geometry2/blob/71b054249cda9f1bbdbdc2f61bd518b998ce0378/tf2/include/tf2/impl/convert.h#L43-L45

sloretz reviewed (approved):
LGTM with a recommendation to remove the templated declarations of `toMsg()` and `fromMsg()` in `tf2/convert.h` since they're no longer being used here.

I'm not sure if this can be backported. If a compiler chooses to ignore `inline`, I think the mangled name on the functions is changed when these become overloads instead of specializations.

jacobperron commented:
@tfoote Before I merge this, maybe you can shed some light on if template specialization is required or if switching to overloads is fine.

jacobperron committed:
Remove last occurance of fromMsg template specialization

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Update tests to use tf2::convert

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
@sloretz I think you're right that this is not eligible for backport, but users can still workaround the issue by explicitly providing template arguments.

I've rebased to get the latest changes and I've removed the last occurrence of the template specialization for `fromMsg` in `tf2_geometry_msgs` (499afbd). I've also updated the tests to use `tf2::convert` now that it is available (143e90d). PTAL

jacobperron commented:
> recommendation to remove the templated declarations of toMsg() and fromMsg() in tf2/convert.h since they're no longer being used here.

Since the templated declarations may be used elsewhere (outside of this repo), it might be better to leave them for now.

jacobperron commented:
Rerunning after the rebase:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=8467)](http://ci.ros2.org/job/ci_linux/8467/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=4358)](http://ci.ros2.org/job/ci_linux-aarch64/4358/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6880)](http://ci.ros2.org/job/ci_osx/6880/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=8368)](http://ci.ros2.org/job/ci_windows/8368/)

nuclearsandwich commented:
@jacobperron removed this from backport consideration since it is not ABI compatible.
